### ISSUE TITLE
Native text selection appears, cursor does not appear if a package operates on the buffer within cursor movement events

### ISSUE THREAD

**Issue by sompylasar** [atom/atom#16267]
I'm developing a package [`atom-cursor-indent`](https://github.com/sompylasar/atom-cursor-indent) which makes the cursor follow the desired (auto) indentation on empty lines.

I've just published the [`0.2.0` version](https://github.com/sompylasar/atom-cursor-indent/releases/tag/v0.2.0) which fixes several issues with the behavior of the package, but the issue reproduces in the [previous version](https://github.com/sompylasar/atom-cursor-indent/releases/tag/v0.1.2) as well.

I'm seeking for an advice, maybe I'm doing something wrong with the Atom objects (the documentation on how to do things is scarce, reading the source to discover methods that do what I need). And I apologize for the source being in CoffeeScript, that was the language most Atom source and packages were in when I started. Thank you in advance.

### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * [ ] Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode – _No, because the issue surfaces only when I'm dealing with the editor from the package I'm developing._
    * [x] Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * [x] Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * [x] Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * [x] Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

There are two issues that I think are closely related:
- At times, when I attempt to select text with a mouse pointer, the native text selection appears (the blue highlight, like in a web browser).
- At times, when I attempt to put a cursor to a certain position with a mouse pointer, the cursor does not appear, and when I type letters afterwards, the text doesn't change.

This doesn't happen every time I select text or put the cursor, but doesn't look like a timing issue from the first glance: regardless of the speed in which I'm doing the selections, it may or may not reproduce.

I noticed that when the cursor is not visible, the `editor` element doesn't have the `is-focused` class name.

What I think affects the behavior is that my package modifies the buffer, so it re-renders while the user drags the mouse, and some internal state related to the DOM gets messed up. This issue may be related: https://github.com/atom/atom/issues/11691

### Steps to Reproduce

1. Install [`atom-cursor-indent`](https://atom.io/packages/atom-cursor-indent)
2. Open a text file in Atom
3. Click around, select text with a mouse.

**Expected behavior:**

The native selection does not appear, only the Atom selection. The cursor is always visible after clicking into the text editor.

**Actual behavior:** 

A picture is worth thousand words, please have a look:
![atom-cursor-indent-selection](https://user-images.githubusercontent.com/498274/33204164-7d814dce-d0b7-11e7-8b01-324ee3379f23.gif)
![atom-cursor-indent-focus](https://user-images.githubusercontent.com/498274/33204163-7d6b47fe-d0b7-11e7-8b95-3ba6daed3e43.gif)

**Reproduces how often:** 

Not measured scientifically, but I'd say ~50%.

### Versions

You can get this information from copy and pasting the output of `atom --version` and `apm --version` from the command line. Also, please include the OS and what version of the OS you're running.

```
> atom --version
Atom    : 1.22.1
Electron: 1.6.15
Chrome  : 56.0.2924.87
Node    : 7.4.0

> apm --version
apm  1.18.8
npm  3.10.10
node 6.9.5 x64
python 2.7.10
git 2.9.3

> uname -a
Darwin ivanbabaks-MacBook-Pro.local 16.0.0 Darwin Kernel Version 16.0.0: Mon Aug 29 17:56:20 PDT 2016; root:xnu-3789.1.32~3/RELEASE_X86_64 x86_64
```

The same is reproduced in Atom Beta, currently `1.23.0-beta1`

### Additional Information

Will provide extra info upon request.

**Comment by cleanenergyjoy** [comment]
majorjammer830@gmail.com
Em 24/11/2017 15:17, "Ivan Babak" <notifications@github.com> escreveu:

**Comment by sompylasar** [comment]
I'm trying to track down the cause of the issue.

When the above happens (when the editor loses focus and the cursor stops blinking after mouse click), in the [following code](https://github.com/atom/atom/blob/v1.22.1/src/text-editor-component.js#L1479-L1486):
- `event.relatedTarget` is `null`
- `this.element` is `atom-text-editor.editor.is-focused`, 
- `event.type` is `"blur"`, 
- `event.currentTarget` is `input.hidden-input`.

```
  didBlurHiddenInput (event) {
    if (this.element !== event.relatedTarget && !this.element.contains(event.relatedTarget)) {
      this.focused = false
      this.stopCursorBlinking()
      this.scheduleUpdate()
      this.element.dispatchEvent(new FocusEvent(event.type, event))
    }
  }
```
<img width="749" alt="screen shot 2017-11-28 at 10 13 28 pm" src="https://user-images.githubusercontent.com/498274/33360816-54a7d310-d48a-11e7-8e8c-f0d394553d64.png">

**Comment by sompylasar** [comment]
So, the cause of the issue is synchronous execution of event handlers during the editor transition from one state to another, and some code affecting that state from within an event handler.

In my case (see comments inline):
```js
    Selection.prototype.markerDidChange = function(e) {
      var cursorMovedEvent, newHeadBufferPosition, newHeadScreenPosition, oldHeadBufferPosition, oldHeadScreenPosition, oldTailBufferPosition, oldTailScreenPosition, textChanged;
      oldHeadBufferPosition = e.oldHeadBufferPosition, oldTailBufferPosition = e.oldTailBufferPosition, newHeadBufferPosition = e.newHeadBufferPosition;
      oldHeadScreenPosition = e.oldHeadScreenPosition, oldTailScreenPosition = e.oldTailScreenPosition, newHeadScreenPosition = e.newHeadScreenPosition;
      textChanged = e.textChanged;
      if (!oldHeadScreenPosition.isEqual(newHeadScreenPosition)) {
        this.cursor.goalColumn = null;
        cursorMovedEvent = {
          oldBufferPosition: oldHeadBufferPosition,
          oldScreenPosition: oldHeadScreenPosition,
          newBufferPosition: newHeadBufferPosition,
          newScreenPosition: newHeadScreenPosition,
          textChanged: textChanged,
          cursor: this.cursor
        };
        this.cursor.emitter.emit('did-change-position', cursorMovedEvent);

// HERE: `cursorMoved` calls `this.emitter.emit('did-change-cursor-position', event);`
// my code handles 'did-change-cursor-position'
// by synchronously changing the buffer and the cursor,
// this happens in the `cursorMoved` call stack.
        this.editor.cursorMoved(cursorMovedEvent);

// HERE: the state of the editor has changed again, 
// including the selection state which is tied to the buffer and the cursor state,
// but the code below didn't complete handling the previous state transition.

      }
      this.emitter.emit('did-change-range');
      return this.editor.selectionRangeChanged({
        oldBufferRange: new Range(oldHeadBufferPosition, oldTailBufferPosition),
        oldScreenRange: new Range(oldHeadScreenPosition, oldTailScreenPosition),
        newBufferRange: this.getBufferRange(),
        newScreenRange: this.getScreenRange(),
        selection: this
      });
    };
```

I couldn't determine what exactly causes the `blur` event though.

For my code, I delayed the editor state changes to `process.nextTick`, and it seems to work, but I risk messing the editor state by applying changes asynchronously.

**Comment by rsese** [comment]
Thanks for sharing these details @sompylasar - I had this on my list to try and reproduce and then followup with you for more details before asking the team to take a look though it looks like you're making some progress in your investigation?

**Comment by sompylasar** [comment]
@rsese

I think I'm done with the investigation of the root cause of the `blur` event because this only happens if the plugin code affects the editor state during an unfinished state transition (in the middle of it, synchronously). I decided to make my plugin apply the changes async, and besides clearly seeing the cursor jump, there's very low risk of applying the changes after the same piece of editor state has also been changed by some other piece of code (race condition).

If anybody makes another plugin which changes the editor buffer from within cursor position or selection change event handlers, this issue will potentially affect that plugin as well.

I'm leaving to your discretion whether to close the issue or to address it later.

**Comment by rsese** [comment]
Thanks @sompylasar :+1:

I was able to reproduce the behavior you described when I installed 0.2.0 so we'll leave this open for now and ask the team to take a look.  I'm glad you have a workaround for the issue and thanks again for sharing your findings :bow:

**Comment by sompylasar** [comment]
Looks like fixed, Atom 1.32.1, atom-cursor-indent 0.3.3.
